internal/graph/memory.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package graph
+
+import (
+	"context"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/internal/descriptor"
+	"oras.land/oras-go/v2/internal/status"
+)
+
+// Memory is a memory based UpEdgeFinder.
+type Memory struct {
+	upEdges sync.Map // map[descriptor.Descriptor]map[descriptor.Descriptor]ocispec.Descriptor
+	indexed sync.Map // map[descriptor.Descriptor]bool
+}
+
+// NewMemory creates a new memory UpEdgeFinder.
+func NewMemory() *Memory {
+	return &Memory{}
+}
+
+// Index indexes up edges for each direct down edge of the given node.
+// There is no data consistency issue as long as deletion is not implemented
+// for the underlying storage.
+func (m *Memory) Index(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) error {
+	downEdges, err := content.DownEdges(ctx, fetcher, node)
+	if err != nil {
+		return err
+	}
+
+	return m.index(ctx, node, downEdges)
+}
+
+// Index indexes up edges for all the down edges of the given node.
+// There is no data consistency issue as long as deletion is not implemented
+// for the underlying storage.
+func (m *Memory) IndexAll(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) error {
+	// track content status
+	tracker := status.NewTracker()
+
+	// prepare pre-handler
+	preHandler := HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) (downEdges []ocispec.Descriptor, err error) {
+		defer func() {
+			if err == nil {
+				// mark the node as done on success
+				done, _ := tracker.TryCommit(desc)
+				close(done)
+			}
+		}()

```

Comment:
nit: it seems can be simplified by removing this `defer` block and add `close(done)` at line 90.